File: app/src/main/java/za/co/gundula/app/arereyeng/ui/AReYengMapActivity.java
package za.co.gundula.app.arereyeng.ui;

import android.Manifest;
import android.content.DialogInterface;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.CursorLoader;
import android.support.v4.content.Loader;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;

import butterknife.BindView;
import butterknife.ButterKnife;
import za.co.gundula.app.arereyeng.utils.Constants;
import za.co.gundula.app.arereyeng.R;
import za.co.gundula.app.arereyeng.data.AreYengContract;
import za.co.gundula.app.arereyeng.model.Agency;

import static za.co.gundula.app.arereyeng.utils.Constants.agency_key;

public class AReYengMapActivity extends AppCompatActivity implements LoaderManager.LoaderCallbacks<Cursor>, OnMapReadyCallback {

    @BindView(R.id.toolbar)
    Toolbar toolbar;

    private GoogleMap mMap;

    Agency agency = null;
    private static final int BUS_STOP_LOADER = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_are_yeng_map);

        ButterKnife.bind(this);

        Bundle bundle = getIntent().getExtras();
        if (bundle != null) {
            agency = bundle.getParcelable(agency_key);
        }


        toolbar.setTitle(getResources().getString(R.string.areyeng_map));
        setSupportActionBar(toolbar);
        //noinspection ConstantConditions
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        SupportMapFragment mapFragment = (SupportMapFragment) getSupportFragmentManager()
                .findFragmentById(R.id.map);
        mapFragment.getMapAsync(this);
        getSupportLoaderManager().initLoader(BUS_STOP_LOADER, null, this);

    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        mMap = googleMap;
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            if (ActivityCompat.shouldShowRequestPermissionRationale(AReYengMapActivity.this, Manifest.permission.ACCESS_FINE_LOCATION)) {
                AlertDialog.Builder builder = new AlertDialog.Builder(this);
                builder.setMessage("Location Permissions")
                        .setCancelable(false)
                        .setPositiveButton("OK", new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int id) {
                                ActivityCompat.requestPermissions(AReYengMapActivity.this, new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, Constants.REQUEST_LOCATION);
                            }
                        });

                AlertDialog alert = builder.create();
                alert.show();

            } else {
                ActivityCompat.requestPermissions(AReYengMapActivity.this, new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, Constants.REQUEST_LOCATION);
            }
        } else {

            mMap.setMyLocationEnabled(true);
            mMap.getUiSettings().setMyLocationButtonEnabled(true);
        }

        mMap.setMapType(GoogleMap.MAP_TYPE_TERRAIN);
        mMap.setBuildingsEnabled(true);
        mMap.setIndoorEnabled(false);
        mMap.getUiSettings().setZoomControlsEnabled(true);
        mMap.getUiSettings().setCompassEnabled(true);
        mMap.animateCamera(CameraUpdateFactory.zoomIn());

    }

    @Override
    public Loader onCreateLoader(int id, Bundle args) {
        Uri bus_stop_uri = AreYengContract.BusStopEntry.CONTENT_URI;

        String[] BUS_STOP_PROJECTION = new String[]{
                AreYengContract.BusStopEntry.COLUMN_ID,
                AreYengContract.BusStopEntry.COLUMN_NAME,
                AreYengContract.BusStopEntry.COLUMN_HREF,
                AreYengContract.BusStopEntry.COLUMN_GEOMETRY_LATITUDE,
                AreYengContract.BusStopEntry.COLUMN_GEOMETRY_LONGITUDE,
                AreYengContract.BusStopEntry.COLUMN_GEOMETRY_TYPE,
        };

        return new CursorLoader(
                this,
                bus_stop_uri,
                BUS_STOP_PROJECTION,
                null,
                null,
                null
        );
    }

    public void onResume() {
        super.onResume();
        getSupportLoaderManager().restartLoader(BUS_STOP_LOADER, null, this);
    }


    @Override
    public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
        if (data != null && data.moveToFirst()) {
            do {

                String station_name = data.getString(1);
                String station_url = data.getString(2);
                double station_latitude = data.getDouble(3);
                double station_longitude = data.getDouble(4);

                addMapMarker(station_latitude, station_longitude, station_name, station_name, 10);

            } while (data.moveToNext());
        }
    }

    @Override
    public void onLoaderReset(Loader loader) {

    }


    public void addMapMarker(double latitude, double longitude, String title, String snippet, int iconResId) {

        mMap.addMarker(new MarkerOptions()
                .position(new LatLng(latitude, longitude))
                .anchor(0.5f, 0.5f)
                .title(title)
                .snippet(snippet));
    }

}

File: app/src/main/java/za/co/gundula/app/arereyeng/ui/AReYengMapActivity.kt
package za.co.gundula.app.arereyeng.ui

import android.Manifest
import android.content.pm.PackageManager
import android.database.Cursor
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.app.ActivityCompat
import androidx.loader.app.LoaderManager
import androidx.loader.content.CursorLoader
import androidx.loader.content.Loader
import butterknife.BindView
import butterknife.ButterKnife
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions
import za.co.gundula.app.arereyeng.R
import za.co.gundula.app.arereyeng.data.AreYengContract
import za.co.gundula.app.arereyeng.model.Agency
import za.co.gundula.app.arereyeng.ui.AReYengMapActivity
import za.co.gundula.app.arereyeng.utils.Constants

class AReYengMapActivity : AppCompatActivity(), LoaderManager.LoaderCallbacks<Cursor?>, OnMapReadyCallback {

    @BindView(R.id.toolbar)
    var toolbar: Toolbar? = null
    private var mMap: GoogleMap? = null
    var agency: Agency? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_are_yeng_map)
        ButterKnife.bind(this)
        val bundle = intent.extras
        if (bundle != null) {
            agency = bundle.getParcelable(Constants.agency_key)
        }
        toolbar!!.title = resources.getString(R.string.areyeng_map)
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        val mapFragment = supportFragmentManager
                .findFragmentById(R.id.map) as SupportMapFragment?
        mapFragment!!.getMapAsync(this)
        supportLoaderManager.initLoader(BUS_STOP_LOADER, null, this)
    }

    override fun onMapReady(googleMap: GoogleMap) {
        mMap = googleMap
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            if (ActivityCompat.shouldShowRequestPermissionRationale(this@AReYengMapActivity, Manifest.permission.ACCESS_FINE_LOCATION)) {
                val builder = AlertDialog.Builder(this)
                builder.setMessage(getString(R.string.location_permissions))
                        .setCancelable(false)
                        .setPositiveButton(getString(R.string.dialog_ok)) { dialog, id -> ActivityCompat.requestPermissions(this@AReYengMapActivity, arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), Constants.REQUEST_LOCATION) }
                val alert = builder.create()
                alert.show()
            } else {
                ActivityCompat.requestPermissions(this@AReYengMapActivity, arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), Constants.REQUEST_LOCATION)
            }
        } else {
            mMap!!.isMyLocationEnabled = true
            mMap!!.uiSettings.isMyLocationButtonEnabled = true
        }
        mMap!!.mapType = GoogleMap.MAP_TYPE_TERRAIN
        mMap!!.isBuildingsEnabled = true
        mMap!!.isIndoorEnabled = false
        mMap!!.uiSettings.isZoomControlsEnabled = true
        mMap!!.uiSettings.isCompassEnabled = true
        mMap!!.animateCamera(CameraUpdateFactory.zoomIn())
    }

    override fun onCreateLoader(id: Int, args: Bundle?): Loader<Cursor?> {
        val bus_stop_uri = AreYengContract.BusStopEntry.CONTENT_URI
        val BUS_STOP_PROJECTION = arrayOf(
                AreYengContract.BusStopEntry.COLUMN_ID,
                AreYengContract.BusStopEntry.COLUMN_NAME,
                AreYengContract.BusStopEntry.COLUMN_HREF,
                AreYengContract.BusStopEntry.COLUMN_GEOMETRY_LATITUDE,
                AreYengContract.BusStopEntry.COLUMN_GEOMETRY_LONGITUDE,
                AreYengContract.BusStopEntry.COLUMN_GEOMETRY_TYPE)
        return CursorLoader(
                this,
                bus_stop_uri,
                BUS_STOP_PROJECTION,
                null,
                null,
                null
        )
    }

    public override fun onResume() {
        super.onResume()
        supportLoaderManager.restartLoader(BUS_STOP_LOADER, null, this)
    }

    override fun onLoadFinished(loader: Loader<Cursor?>, data: Cursor?) {
        if (data != null && data.moveToFirst()) {
            do {
                val station_name = data.getString(1)
                val station_url = data.getString(2)
                val station_latitude = data.getDouble(3)
                val station_longitude = data.getDouble(4)
                addMapMarker(station_latitude, station_longitude, station_name, station_name, 10)
            } while (data.moveToNext())
        }
    }

    override fun onLoaderReset(loader: Loader<*>?) {}
    fun addMapMarker(latitude: Double, longitude: Double, title: String?, snippet: String?, iconResId: Int) {
        mMap!!.addMarker(MarkerOptions()
                .position(LatLng(latitude, longitude))
                .anchor(0.5f, 0.5f)
                .title(title)
                .snippet(snippet))
    }

    companion object {
        private const val BUS_STOP_LOADER = 0
    }
}